### PR TITLE
[Model]  Adding Granite MoE.

### PR DESCRIPTION
**Pull request by shawntan**
Adds Granite MoE class to vLLM. 
Model will be on Hugging Face once https://github.com/huggingface/transformers/pull/33207  is merged.

### CODE REVIEW

File: tests/models/test_granitemoe.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+def test_models(
+    hf_runner,
+    vllm_runner,
+    example_prompts,
+    model: str,
+    dtype: str,
+    max_tokens: int,
+    num_logprobs: int,
+) -> None:
+    # TODO(sang): Sliding window should be tested separately.
```
Comment (mgoin) [inline tests/models/test_granitemoe.py:36]: does this model have sliding window?

File: vllm/model_executor/models/granitemoe.py [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+                                               ReplicatedLinear,
+                                               RowParallelLinear)
+from vllm.model_executor.layers.logits_processor import LogitsProcessor
+from vllm.model_executor.layers.quantization.base_config import (
+    QuantizationConfig)
+from vllm.model_executor.layers.rotary_embedding import get_rope
+from vllm.model_executor.layers.vocab_parallel_embedding import (
+    DEFAULT_VOCAB_PADDING_SIZE, ParallelLMHead, VocabParallelEmbedding)
+from vllm.model_executor.sampling_metadata import SamplingMetadata
+# from vllm.sequence import IntermediateTensors, SamplerOutput
```
Comment (mgoin) [inline vllm/model_executor/models/granitemoe.py:46]: cruft?

File: vllm/model_executor/models/granitemoe.py [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Inference-only GraniteMoe model."""
+from typing import Iterable, List, Optional, Tuple
+
+import torch
+from torch import nn
+# from transformers import GraniteMoeConfig
```
Comment (mgoin) [inline vllm/model_executor/models/granitemoe.py:28]: cruft?

**Review by mgoin — COMMENTED** [review]
Could you describe how this is different from mixtral? It seems fairly similar at a glance. Is it just extra config details or names?

File: vllm/model_executor/models/granitemoe.py (hunk 3 above)
Comment (shawntan) [reply vllm/model_executor/models/granitemoe.py:28]: I can remove it, but it is left there while the huggingface PR is not accepted.

**Comment by shawntan** [comment]
> Could you describe how this is different from mixtral? It seems fairly similar at a glance. Is it just extra config details or names?

We have additional config parameters for the muP settings, the loading script is different because of the layout of the checkpoint on huggingface, and we eventually will update the class to have shared experts.

**Comment by DarkLight1337** [comment]
A quick heads-up that the new locations of the model tests have been adjusted in #7820, so please merge from main.

**Comment by shawntan** [comment]
https://github.com/huggingface/transformers/pull/33207 merged.

**Comment by shawntan** [comment]
@njhill could this be reviewed?

**Comment by njhill** [comment]
Will test this out after we update to the transformers release containing the `GraniteMoeConfig` class, hopefully happening today!

**Comment by shawntan** [comment]
Updated both Granite and GraniteMoE to use the Transformer config files.

**Comment by shawntan** [comment]
Not sure if this still works, I don't know the best way to undo a rebase.

File: vllm/model_executor/models/granitemoe.py [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+        hidden_states = self.model(input_ids, positions, kv_caches,
+                                   attn_metadata, intermediate_tensors)
+        return hidden_states
+
+    def compute_logits(self, hidden_states: torch.Tensor,
+                       sampling_metadata: SamplingMetadata) -> torch.Tensor:
+        logits = self.logits_processor(self.lm_head, hidden_states,
+                                       sampling_metadata)
+        logits = logits / self.config.logits_scaling
+        return logits
```
Comment (njhill) [inline vllm/model_executor/models/granitemoe.py:389]: This is required for tensor parallel
```suggestion
                       sampling_metadata: SamplingMetadata) -> Optional[torch.Tensor]:
        logits = self.logits_processor(self.lm_head, hidden_states,
                                       sampling_metadata)
        if logits is not None:
            logits /= self.config.logits_scaling
        return logits
```

File: vllm/model_executor/models/granitemoe.py [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+            assert self.total_num_kv_heads % tp_size == 0
+        else:
+            # Number of KV heads is less than TP size, so we replicate
+            # the KV heads across multiple tensor parallel GPUs.
+            assert tp_size % self.total_num_kv_heads == 0
+        self.num_kv_heads = max(1, self.total_num_kv_heads // tp_size)
+        self.head_dim = hidden_size // self.total_num_heads
+        self.q_size = self.num_heads * self.head_dim
+        self.kv_size = self.num_kv_heads * self.head_dim
+        # self.scaling = self.head_dim**-0.5
```
Comment (njhill) [inline vllm/model_executor/models/granitemoe.py:135]: Remove this line?

File: vllm/model_executor/models/granitemoe.py [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+    ) -> torch.Tensor:
+        hidden_states = self.model(input_ids, positions, kv_caches,
+                                   attn_metadata, intermediate_tensors)
+        return hidden_states
+
+    def compute_logits(self, hidden_states: torch.Tensor,
+                       sampling_metadata: SamplingMetadata) -> torch.Tensor:
+        logits = self.logits_processor(self.lm_head, hidden_states,
+                                       sampling_metadata)
+        logits = logits / self.config.logits_scaling
```
Comment (njhill) [inline vllm/model_executor/models/granitemoe.py:388]: @shawntan instead of applying the scaling factor here, could the reciprocal be passed to the `LogitsProcessor` constructor as the `scale` parameter?

The difference is whether it gets applies before or after the any custom logits processors that might be tied to particular requests. Do you know what the behaviour is in the transfomers implementation in this regard?

The same question applies actually to the regular granite impl [here](https://github.com/vllm-project/vllm/blob/c5d55356f9d2b2075ac53cf20453358c1e2b7bde/vllm/model_executor/models/granite.py#L406).

**Review by njhill — COMMENTED** [review]
@shawntan I did another pass, just a few final comments.

**Review by njhill — APPROVED** [review]
Thanks @shawntan

File: vllm/model_executor/models/__init__.py [hunk 7]
```diff
@@ -71,6 +71,9 @@
     "MedusaModel": ("medusa", "Medusa"),
     "EAGLEModel": ("eagle", "EAGLE"),
     "MLPSpeculatorPreTrainedModel": ("mlp_speculator", "MLPSpeculator"),
+    "JambaForCausalLM": ("jamba", "JambaForCausalLM"),
+    "GraniteForCausalLM": ("granite", "GraniteForCausalLM"),
+    "GraniteMoeForCausalLM": ("granitemoe", "GraniteMoeForCausalLM")
```
Comment (DarkLight1337) [inline vllm/model_executor/models/__init__.py:76]: Please move the entries (except for the existing ones under NOTE) to maintain their alphabetical order. You should also update the Supported Models page to include this model.

**Comment by njhill** [comment]
@DarkLight1337 I have opened a separate PR for the docs to avoid having to rebuild this one :) https://github.com/vllm-project/vllm/pull/9025 .. also updated the non-MoE granite doc in that.

So this one should hopefully be good to merge...

**Comment by DarkLight1337** [comment]
Sure, let's merge this.

**Comment by njhill** [comment]
Thanks @shawntan @DarkLight1337
